pkg/operator/ceph/cluster/osd/health.go

```diff
@@ -90,29 +82,60 @@ func (m *Monitor) osdStatus() error {
 		id := int(id64)
 
 		logger.Debugf("validating status of osd.%d", id)
-		_, tracked := m.lastStatus[id]
 
-		// No action on in/out cluster state is taken at this time.
-		status, _, err := osdDump.StatusByID(int64(id))
+		status, in, err := osdDump.StatusByID(int64(id))
 		if err != nil {
 			return err
 		}
 
 		if status != upStatus {
 			logger.Infof("osd.%d is marked 'DOWN'", id)
-			if tracked {
-				evalDownStatus(id)
-			} else {
-				m.lastStatus[id] = time.Now()
-			}
 		} else {
 			logger.Debugf("osd.%d is healthy.", id)
-			if tracked {
-				logger.Infof("osd.%d recovered, stopping tracking.", id)
-				delete(m.lastStatus, id)
-			}
+
+		}
+
+		if in != inStatus {
+			logger.Infof("osd.%d is marked 'OUT'", id)
+			m.outOSDs[strconv.Itoa(id)] = true
+		} else {
+			m.outOSDs[strconv.Itoa(id)] = false
 		}
 	}
+	if err := m.handleOSDMarkedOut(m.outOSDs); err != nil {
+		return err
+	}
+	return nil
+}
 
+func (m *Monitor) handleOSDMarkedOut(outOSDs map[string]bool) error {
+	if len(outOSDs) == 0 {
+		return nil
+	}
+
+	var outOSDids []string
+	for osdID, outStatus := range outOSDs {
+		if outStatus {
+			outOSDids = append(outOSDids, osdID)
+		}
+	}
+
+	safeToDestroyOSDs, err := client.OsdSafeToDestroy(m.context, m.clusterName, outOSDids)
+	if err != nil {
+		return err
+	}
+
+	for _, id := range safeToDestroyOSDs.SafeToDestroy {
+		logger.Infof("osd.%d is 'safe-to-destroy'", id)
+		label := fmt.Sprintf("ceph-osd-id=%d", id)
+		dp, err := k8sutil.GetDeployments(m.context.Clientset, m.clusterName, label)
+		if err != nil {
```

Comment:
A valid error could be that the deployment is not found. In fact, we would expect this frequently. After an OSD has been removed by this code path previously, the OSD will look like it needs to be removed every time this check is done. So we will need to ignore the error in that case.

Reply:
Got it.